gadget/gadget.go

```diff
@@ -368,9 +383,10 @@ func validateVolume(name string, vol *Volume) error {
 	// for validating structure overlap
 	structures := make([]LaidOutStructure, len(vol.Structure))
 
+	state := &validationState{}
 	previousEnd := Size(0)
 	for idx, s := range vol.Structure {
-		if err := validateVolumeStructure(&s, vol); err != nil {
+		if err := validateVolumeStructure(&s, vol, constraints, state); err != nil {
```

Comment:
I still think we can keep the consistency checks only at the volume level, without the need for passing the state down the stack. In the end the only information that is collected is whether a system-seed role was seen or not and we are also going through volume structures one by one.

How about we relax the rules in validateRole to not verify the label, but only check the role, i.e. whether it's one of the allowed patterns, and whether MBR makes any sense. Then refactor the code to something like this:

```
const (
	...
	// ImplicitSystemDataLabel is the implicit filesystem label of structure
	// of system-data role
	ImplicitSystemDataLabel = "writable"
	// UbuntuSystemDataLabel is the expected label of system data structure
	// when the system uses a recovery & seeding mode
	UbuntuSystemDataLabel   = "ubuntu-data"
)

type validationState struct {
	SystemSeed      *VolumeStructure
	SystemData      *VolumeStructure
}

func validateVolume(name string, vol *Volume, constraints *ModelConstraints) error {
	...
	state := &validationState{}
	previousEnd := Size(0)
	for idx, s := range vol.Structure {
		if err := validateVolumeStructure(&s, vol); err != nil {
			return fmt.Errorf("invalid structure %v: %v", fmtIndexAndName(idx, s.Name), err)
		}
		...
		switch s.Role {
		case SystemSeed:
			if state.SystemSeed != nil {
				return fmt.Errorf("cannot have more than one system-seed structure")
			}
			state.SystemSeed = &vol.Structure[idx]
		case SystemData:
			if state.SystemData != nil {
				return fmt.Errorf("cannot have more than one system-data structure")
			}
			state.SystemData = &vol.Structure[idx]
		}

		previousEnd = end
	}

	// gadget must be auto-consistent if constraints are not specified
	if err := ensureVolumeConsistency(state, constraints); err != nil {
		return err
	}
	...
}

func ensureVolumeConsistency(state *validationState, constraints *ModelConstraints) error {
	if constraints == nil {
		if state.SystemData != nil {
			allowedSystemDataLabel := ImplicitSystemDataLabel
			if state.SystemSeed != nil {
				allowedSystemDataLabel = UbuntuSystemDataLabel
			}
			if state.SystemData.Label != "" && state.SystemData.Label != allowedSystemDataLabel {
				return fmt.Errorf("system-data structure must have an implicit label or %q, not %q",
					allowedSystemDataLabel, state.SystemData.Label)
			}
		}

		return nil
	}

	if state.SystemSeed != nil {
		if !constraints.SystemSeed {
			return fmt.Errorf("model does not support system-seed role")
		}
		if state.SystemData != nil && state.SystemData.Label != UbuntuSystemDataLabel {
			return fmt.Errorf("system-data structure must have a label %q, not %q",
				UbuntuSystemDataLabel, state.SystemData.Label)
		}
	} else {
		// error if we want system-seed constraint but no actual system-seed structure
		if constraints.SystemSeed {
			return fmt.Errorf("model requires system-seed structure, but none was found")
		}
	}

	return nil
}


// simplify 'detached' role validation
func validateRole(vs *VolumeStructure, vol *Volume) error {
	...
	switch vsRole {
	case SystemData, SystemSeed:
		// roles have cross dependencies, consistency checks are done at
		// the volume level
	case MBR:
		if vs.Size > SizeMBR {
			return errors.New("mbr structures cannot be larger than 446 bytes")
		}
		if vs.Offset != nil && *vs.Offset != 0 {
			return errors.New("mbr structure must start at offset 0")
		}
		if vs.ID != "" {
			return errors.New("mbr structure must not specify partition ID")
		}
		if vs.Filesystem != "" && vs.Filesystem != "none" {
			return errors.New("mbr structures must not specify a file system")
		}
	case SystemBoot, BootImage, BootSelect, "":
		// noop
	default:
		return fmt.Errorf("unsupported role")
	}
	return nil
}
```

Reply:
this refactor proposed by @bboozzoo looks reasonable to me (I might be missing something)

Reply:
This looks good. Will refactor as suggested.